W-13104970: Add instructions for working with code snippets (#49)
mulesoft/docs-code-builder: modules/ROOT/pages/work-with-code-snippets.adoc

= Work with Code Snippets

> Code snippets are reusable templates that make it easier to enter repeating code patterns,  such as an insert operation from Database Connector, or logging messages and variables.

Reviewer: ```suggestion
Code snippets are reusable templates that simplify reusing code patterns, such as an insert operation from Database Connector, or logging messages and variables.
```

> Anypoint Code Builder supports both built-in snippets within the web IDE, and enables you to create and save your own snippets.

Reviewer: ```suggestion
Anypoint Code Builder provides built-in snippets and enables you to create and save custom snippets.
```
Please confirm if this is the intended meaning.

> Code Builder does not filter  document that we don't filter out config snippets, despite the fact that you cannot use them in the canvas.

Reviewer: Unclear sentence, please rewrite.

== Built-In Snippets

Code Builder bundles a set of built-in snippets. To add a built-in snippet to your flow:

. Select the `plus` icon on your canvas to add a new component to your flow.
. Select _Snippets_:
+
image::select-snippet-overview.png[]
. Select _Built-in Snippets_:
+
image::built-in-snippets-overview.png[]
. Choose the snippet you want from the list of provided snippets.

> Additionally, you can use any of the provided snippets by using the auto-complete feature when working on the XML editor.

Reviewer: ```suggestion
Use the auto-complete feature when working on the XML editor to find any of the built-in snippets.
```

> == Create Your Own Snippet

Reviewer: ```suggestion
== Create a Snippet
```

> You can create your own reusable snippets and save them to reuse them in your project.

Reviewer: ```suggestion
Create and save reusable custom snippets to use in your project.
```

> The snippets that you save are only available to you. Code Builder does not support shared snippets with all users working in a project.

Reviewer: ```suggestion
The snippets that you save are not available to others. Code Builder does not support sharing snippets with all users working on a project.
```

> To create your own snippets:

Reviewer: ```suggestion
To create a snippet:
```

. Select the `plus` icon on your canvas to add a new component to your flow.
. Select _Snippets_:
+
image::select-snippet-overview.png[]
. Select _User Snippets_:
+
image::user-snippet-overview.png[]
. Select the `plus` icon:
+
image::create-user-snippet-overview.png[]
+
If you don't have other user snippets in your web IDE, Code Builder prompts you to create one.

> . Code Builder creates a new `mule-xml.json`  snippet file with a sample snippet for you to edit.

Reviewer: ```suggestion
+ 
Code Builder creates a new `mule-xml.json` snippet file with a sample snippet for you to edit.
```
This is more a clarification than a step.

> . Uncomment all the fields after *Example*:

Reviewer: ```suggestion
. In the sample snippet file (`mule-xml.json`) Code Builder creates, uncomment all the fields after *Example*:
```
Or, you could move the previous section to this step.

+
[source,JSON]
--
{
    // Example:
	 "Log a message": { //<1>
	 	"prefix": "log", //<2>
	 	"body": [ //<3>
	 		"<logger level=\"${1:INFO}\" message='$2' doc:name=\"Logger\" doc:id=\"$RANDOM_HEX\" />$0",
	 	],
	 	"description": "Log output to console" //<4>
	 }
}
--

> <1> The first element in the XML is the text that will display in the list of user snippets.

Reviewer: ```suggestion
<1> The first element in the XML is the text that displays in the list of user snippets.
```

> <2> The `prefix` field is the trigger text for that snippet. You can start typing it and press `tab` in your XML editor to add the snippet using auto-complete.

Reviewer: ```suggestion
<2> The `prefix` field is the trigger text for that snippet. Start typing the prefix and press `tab` in your XML editor to add the snippet using auto-complete.
```

<3> The `body` field is the actual code template that you can reuse.
<4> The `description` field is a description of the snippet.
. You can add other snippets to your `mule-xml.json` by adding new fields. For example:
+
[source,JSON]
--
{
    // Example:
	 "Log a message": {
	 	"prefix": "log",
	 	"body": [
	 		"<logger level=\"${1:INFO}\" message='$2' doc:name=\"Logger\" doc:id=\"$RANDOM_HEX\" />$0",
	 	],
	 	"description": "Log output to console"
	 },
        "Select first and last name from database": { //<1>
		"prefix": "select-name-from-db",
		"body": [
          "<db:select config-ref=\"${1:Database_Config}\" doc:name=\"${2:Select Contact from DB}\" doc:id=\"$RANDOM_HEX\" >",
          "<db:sql>",
          "<![CDATA[SELECT * FROM Contacts WHERE FirstName = :firstname AND LastName = :lastname]]>",
          "</db:sql>",
          "<db:input-parameters>",
            "<![CDATA[#[\{",
            "firstname: vars.firstnameVar,",
            "lastname: vars.lastnameVar",
            "\\}]]]>",
          "</db:input-parameters>",
          "</db:select>$0"
		],
		"description": "Pull first and last name from database"
	 }
}
--

> <1> The previous sample shows a snippet for a Database Connector select operation that retrieves a `firstName` and `lastName` values from a Database.

Reviewer: ```suggestion
<1> The previous sample shows a snippet for a Database Connector select operation that retrieves `firstName` and `lastName` values from a Database.
```
